# Wrangle Pisa 2012
#### By Gabriela Sikora


### Abstract



### Table of Contents

<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
    <ul>
    <li><a href="#gathering">Gathering Data</a></li>
    <li><a href="#assessing1">Initial Assessment</a></li>
        <ul>
          <li><a href="#issues1">Identified Issues</a></li>
        </ul>
    <li><a href="#cleaning1">Cleaning Data</a></li>
    <li><a href="#assessing2">Final Assessment</a></li>
    </ul>
<li><a href="#sav">Store</a></li>
<li><a href="#conclusion">Conclusion</a></li>
<li><a href="#references">References</a></li>    
</ul>

<a id='intro'></a>
## Introduction



To begin, we first need to import all the relevant libraries:
#### Import

In [83]:
# import all packages
import numpy as np
import pandas as pd

<a id='wrangling'></a>
## Data Wrangling

In this section of the report, we will load in the data, check it for cleanliness, and then trim and clean the datasets for analysis. 

<a id='gathering'></a>
### Gathering Data

Unfortunately the Pisa dataset is too large to be uploaded to github, but if you would like to work with it yourself then place follow this link: https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisa2012.csv.zip 

In [84]:
# Encoding help found here: https://stackoverflow.com/questions/30462807/encoding-error-in-panda-read-csv

# Load in the Pisa dataframe
pisa = pd.read_csv('pisa2012.csv', encoding="cp1252", sep=',')

# Load in the Pisa dictionary
pisa_dict = pd.read_csv('pisadict2012.csv', encoding="cp1252", sep=',')

<a id='assessing1'></a>
### First Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [85]:
pisa

Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD         NC  SCHOOLID  \
0                1  Albania     80000  ALB0006  Non-OECD    Albania         1   
1                2  Albania     80000  ALB0006  Non-OECD    Albania         1   
2                3  Albania     80000  ALB0006  Non-OECD    Albania         1   
3                4  Albania     80000  ALB0006  Non-OECD    Albania         1   
4                5  Albania     80000  ALB0006  Non-OECD    Albania         1   
5                6  Albania     80000  ALB0006  Non-OECD    Albania         1   
6                7  Albania     80000  ALB0006  Non-OECD    Albania         1   
7                8  Albania     80000  ALB0006  Non-OECD    Albania         1   
8                9  Albania     80000  ALB0006  Non-OECD    Albania         1   
9               10  Albania     80000  ALB0005  Non-OECD    Albania         2   
10              11  Albania     80000  ALB0005  Non-OECD    Albania         2   
11              12  Albania     80000  ALB0005  Non-OECD    Albania         2   
12              13  Albania     80000  ALB0005  Non-OECD    Albania         2   
13              14  Albania     80000  ALB0005  Non-OECD    Albania         2   
14              15  Albania     80000  ALB0005  Non-OECD    Albania         2   
15              16  Albania     80000  ALB0005  Non-OECD    Albania         2   
16              17  Albania     80000  ALB0005  Non-OECD    Albania         2   
17              18  Albania     80000  ALB0005  Non-OECD    Albania         2   
18              19  Albania     80000  ALB0005  Non-OECD    Albania         2   
19              20  Albania     80000  ALB0005  Non-OECD    Albania         2   
20              21  Albania     80000  ALB0005  Non-OECD    Albania         2   
21              22  Albania     80000  ALB0005  Non-OECD    Albania         2   
22              23  Albania     80000  ALB0005  Non-OECD    Albania         2   
23              24  Albania     80000  ALB0005  Non-OECD    Albania         2   
24              25  Albania     80000  ALB0005  Non-OECD    Albania         2   
25              26  Albania     80000  ALB0005  Non-OECD    Albania         2   
26              27  Albania     80000  ALB0005  Non-OECD    Albania         2   
27              28  Albania     80000  ALB0005  Non-OECD    Albania         2   
28              29  Albania     80000  ALB0005  Non-OECD    Albania         2   
29              30  Albania     80000  ALB0005  Non-OECD    Albania         2   
...            ...      ...       ...      ...       ...        ...       ...   
485460      485461  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485461      485462  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485462      485463  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485463      485464  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485464      485465  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485465      485466  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485466      485467  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485467      485468  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485468      485469  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485469      485470  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485470      485471  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485471      485472  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485472      485473  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485473      485474  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485474      485475  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485475      485476  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485476      485477  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485477      485478  Vietnam   7040000  

In [86]:
pisa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Columns: 636 entries, Unnamed: 0 to VER_STU
dtypes: float64(250), int64(18), object(368)
memory usage: 2.3+ GB


In [87]:
pisa.dtypes[400:]

ST23Q08        object
ST24Q01        object
ST24Q02        object
ST24Q03        object
CLCUSE1        object
CLCUSE301       int64
CLCUSE302       int64
DEFFORT         int64
QUESTID        object
BOOKID         object
EASY           object
AGE           float64
GRADE         float64
PROGN          object
ANXMAT        float64
ATSCHL        float64
ATTLNACT      float64
BELONG        float64
BFMJ2         float64
BMMJ1         float64
CLSMAN        float64
COBN_F         object
COBN_M         object
COBN_S         object
COGACT        float64
CULTDIST      float64
CULTPOS       float64
DISCLIMA      float64
ENTUSE        float64
ESCS          float64
               ...   
W_FSTR55      float64
W_FSTR56      float64
W_FSTR57      float64
W_FSTR58      float64
W_FSTR59      float64
W_FSTR60      float64
W_FSTR61      float64
W_FSTR62      float64
W_FSTR63      float64
W_FSTR64      float64
W_FSTR65      float64
W_FSTR66      float64
W_FSTR67      float64
W_FSTR68      float64
W_FSTR69  

In [88]:
pisa_dict

Unnamed: 0                                                  x
0           CNT                           Country code 3-character
1      SUBNATIO  Adjudicated sub-region code 7-digit code (3-di...
2       STRATUM  Stratum ID 7-character (cnt + region ID + orig...
3          OECD                                       OECD country
4            NC                       National Centre 6-digit Code
5      SCHOOLID  School ID 7-digit (region ID + stratum ID + 3-...
6       STIDSTD                                         Student ID
7       ST01Q01                                International Grade
8       ST02Q01                           National Study Programme
9       ST03Q01                                      Birth - Month
10      ST03Q02                                        Birth -Year
11      ST04Q01                                             Gender
12      ST05Q01                                   Attend <ISCED 0>
13      ST06Q01                                   Age at <ISCED 1>
14      ST07Q01                                 Repeat - <ISCED 1>
15      ST07Q02                                 Repeat - <ISCED 2>
16      ST07Q03                                 Repeat - <ISCED 3>
17      ST08Q01                          Truancy - Late for School
18      ST09Q01                    Truancy - Skip whole school day
19     ST115Q01           Truancy - Skip classes within school day
20      ST11Q01                                   At Home - Mother
21      ST11Q02                                   At Home - Father
22      ST11Q03                                 At Home - Brothers
23      ST11Q04                                  At Home - Sisters
24      ST11Q05                             At Home - Grandparents
25      ST11Q06                                   At Home - Others
26      ST13Q01                          Mother<Highest Schooling>
27      ST14Q01            Mother Qualifications - <ISCED level 6>
28      ST14Q02           Mother Qualifications - <ISCED level 5A>
29      ST14Q03           Mother Qualifications - <ISCED level 5B>
..          ...                                                ...
605    W_FSTR55           FINAL STUDENT REPLICATE BRR-FAY WEIGHT55
606    W_FSTR56           FINAL STUDENT REPLICATE BRR-FAY WEIGHT56
607    W_FSTR57           FINAL STUDENT REPLICATE BRR-FAY WEIGHT57
608    W_FSTR58           FINAL STUDENT REPLICATE BRR-FAY WEIGHT58
609    W_FSTR59           FINAL STUDENT REPLICATE BRR-FAY WEIGHT59
610    W_FSTR60           FINAL STUDENT REPLICATE BRR-FAY WEIGHT60
611    W_FSTR61           FINAL STUDENT REPLICATE BRR-FAY WEIGHT61
612    W_FSTR62           FINAL STUDENT REPLICATE BRR-FAY WEIGHT62
613    W_FSTR63           FINAL STUDENT REPLICATE BRR-FAY WEIGHT63
614    W_FSTR64           FINAL STUDENT REPLICATE BRR-FAY WEIGHT64
615    W_FSTR65           FINAL STUDENT REPLICATE BRR-FAY WEIGHT65
616    W_FSTR66           FINAL STUDENT REPLICATE BRR-FAY WEIGHT66
617    W_FSTR67           FINAL STUDENT REPLICATE BRR-FAY WEIGHT67
618    W_FSTR68           FINAL STUDENT REPLICATE BRR-FAY WEIGHT68
619    W_FSTR69           FINAL STUDENT REPLICATE BRR-FAY WEIGHT69
620    W_FSTR70           FINAL STUDENT REPLICATE BRR-FAY WEIGHT70
621    W_FSTR71           FINAL STUDENT REPLICATE BRR-FAY WEIGHT71
622    W_FSTR72           FINAL STUDENT REPLICATE BRR-FAY WEIGHT72
623    W_FSTR73           FINAL STUDENT REPLICATE BRR-FAY WEIGHT73
624    W_FSTR74           FINAL STUDENT REPLICATE BRR-FAY WEIGHT74
625    W_FSTR75           FINAL STUDENT REPLICATE BRR-FAY WEIGHT75
626    W_FSTR76           FINAL STUDENT REPLICATE BRR-FAY WEIGHT76
627    W_FSTR77           FINAL STUDENT REPLICATE BRR-FAY WEIGHT77
628    W_FSTR78           FINAL STUDENT REPLICATE BRR-FAY WEIGHT78
629    W_FSTR79           FINAL STUDENT REPLICATE BRR-FAY WEIGHT79
630    W_FSTR80           FINAL STUDENT REPLICATE BRR-FAY WEIGHT80
631    WVARSTRR           RANDOMIZED FINAL VARIANCE STRATUM (1-80)
632    VAR_UNIT                    RANDOMLY ASSIGNED

In [89]:
# Check for null values in relevant columns
columns = ['STIDSTD', 'ST01Q01', 'ST04Q01', 'ST13Q01', 
            'ST17Q01', 'ST28Q01', 'ST57Q01', 'ST57Q02', 'ST57Q03', 
            'ST57Q04', 'ST57Q05', 'MMINS', 'LMINS', 'SMINS',
            'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',  
            'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
            'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']

for x in columns:
    print(x, pisa[x].isna().sum())

STIDSTD 0
ST01Q01 0
ST04Q01 0
ST13Q01 27511
ST17Q01 42229
ST28Q01 11725
ST57Q01 184123
ST57Q02 215682
ST57Q03 201677
ST57Q04 205833
ST57Q05 195988
MMINS 202187
LMINS 202624
SMINS 214576
PV1MATH 0
PV2MATH 0
PV3MATH 0
PV4MATH 0
PV5MATH 0
PV1READ 0
PV2READ 0
PV3READ 0
PV4READ 0
PV5READ 0
PV1SCIE 0
PV2SCIE 0
PV3SCIE 0
PV4SCIE 0
PV5SCIE 0


In [90]:
pisa['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
Australia                   14481
United Kingdom              12659
United Arab Emirates        11500
Switzerland                 11229
Qatar                       10966
Colombia                     9073
Finland                      8829
Belgium                      8597
Denmark                      7481
Jordan                       7038
Chile                        6856
Thailand                     6606
Japan                        6351
Chinese Taipei               6046
Peru                         6035
Slovenia                     5911
Argentina                    5908
Kazakhstan                   5808
Portugal                     5722
Indonesia                    5622
Singapore                    5546
Macao-China                  5335
Czech Republic               5327
Uruguay                      5315
Bulgaria      

In [91]:
pisa.duplicated().sum()

0

<a id='issues1'></a>
### Identified Issues
Here we can see some of the issues noticed in the dataframes. These quality issues will be cleaned in the following section.

1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed
2. Make the variable names readable since they are codes that can only be read with the dictionary
3. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary
4. Missing values in relevant columns ('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')
5. Create a total for Out-of-School Study Time and Learning time (minutes per week)
6. Single countries have various entries in the country column (Florida (USA), Massachusetts (USA), Connecticut (USA) and United States of America)


<a id='cleaning1'></a>
### First Cleaning

In [92]:
# Create a copy to preserve the original dataframe
pisa_clean = pisa.copy()

### Narrow down the dataframe by only looking at variables of interest

There are 636 columns available in the original dataframe, which is far more than what is necessary. We will rather look at just the following variables.

In [93]:
pisa_clean = pisa_clean[['CNT', 'STIDSTD', 'ST04Q01', 'ST13Q01', 'ST14Q01', 'ST14Q02', 'ST14Q03', 'ST14Q04', 
                         'ST17Q01', 'ST18Q01', 'ST18Q02', 'ST18Q03', 'ST18Q04', 
                         'ST57Q01', 'ST57Q02', 'ST57Q03', 'ST57Q04', 'ST57Q05', 'MMINS', 
                         'LMINS', 'SMINS', 'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 
                         'PV5MATH', 'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
                         'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']]       

In [94]:
# Check to see if the columns are correctly reduced
list(pisa_clean)

['CNT',
 'STIDSTD',
 'ST04Q01',
 'ST13Q01',
 'ST14Q01',
 'ST14Q02',
 'ST14Q03',
 'ST14Q04',
 'ST17Q01',
 'ST18Q01',
 'ST18Q02',
 'ST18Q03',
 'ST18Q04',
 'ST57Q01',
 'ST57Q02',
 'ST57Q03',
 'ST57Q04',
 'ST57Q05',
 'MMINS',
 'LMINS',
 'SMINS',
 'PV1MATH',
 'PV2MATH',
 'PV3MATH',
 'PV4MATH',
 'PV5MATH',
 'PV1READ',
 'PV2READ',
 'PV3READ',
 'PV4READ',
 'PV5READ',
 'PV1SCIE',
 'PV2SCIE',
 'PV3SCIE',
 'PV4SCIE',
 'PV5SCIE']

### Create an average of all Math, Reading, and Science scores

For each of the Math, Reading, and Science scores, there are 5 scores provided. The variation between these scores is not needed, so rather we will look at the average for each of these major categories and remove the individual scores.

In [95]:
# Create average of Math scores
pisa_clean['Average Math Score']    = (pisa_clean['PV1MATH'] + pisa_clean['PV2MATH'] + pisa_clean['PV3MATH'] + pisa_clean['PV4MATH'] + pisa_clean['PV5MATH']) / 5

# Create average of Reading scores
pisa_clean['Average Reading Score'] = (pisa_clean['PV1READ'] + pisa_clean['PV2READ'] + pisa_clean['PV3READ'] + pisa_clean['PV4READ'] + pisa_clean['PV5READ']) / 5

# Create average of Science scores
pisa_clean['Average Science Score'] = (pisa_clean['PV1SCIE'] + pisa_clean['PV2SCIE'] + pisa_clean['PV3SCIE'] + pisa_clean['PV4SCIE'] + pisa_clean['PV5SCIE']) / 5

# Create average score
pisa_clean['Average Total Score']   = (pisa_clean['Average Math Score'] + pisa_clean['Average Reading Score'] + pisa_clean['Average Science Score']) / 3

In [96]:
# Remove the individual scores for Math, Reading, and Science
pisa_clean.drop(columns=['PV1MATH',  'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',   
                         'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ',  'PV5READ',
                         'PV1SCIE',  'PV2SCIE', 'PV3SCIE',  'PV4SCIE', 'PV5SCIE'], inplace=True)

In [97]:
# Test to see accuracy of the Math score
print(pisa_clean['Average Math Score'].describe())
print('-'*22)

# Test to see accuracy of the Reading score
print(pisa_clean['Average Reading Score'].describe())
print('-'*22)

# Test to see accuracy of the Science score
print(pisa_clean['Average Science Score'].describe())

# Test to see accuracy of the Total score
print(pisa_clean['Average Total Score'].describe())

count    485490.000000
mean        469.651234
std         100.786610
min          54.767080
25%         396.019620
50%         465.734520
75%         540.123060
max         903.107960
Name: Average Math Score, dtype: float64
----------------------
count    485490.000000
mean        472.006964
std          98.863310
min           6.445400
25%         405.044200
50%         475.477980
75%         542.831195
max         849.359740
Name: Average Reading Score, dtype: float64
----------------------
count    485490.000000
mean        475.808094
std          97.998470
min          25.158540
25%         405.762800
50%         475.512860
75%         546.381920
max         857.832900
Name: Average Science Score, dtype: float64
count    485490.000000
mean        472.488764
std          96.036271
min          77.114593
25%         403.992595
50%         472.046460
75%         541.455700
max         826.592027
Name: Average Total Score, dtype: float64


In [98]:
pisa_clean.sample(10)

CNT  STIDSTD ST04Q01  \
15273   United Arab Emirates    10531  Female   
313716           Macao-China     3349    Male   
371898           New Zealand     3303  Female   
55903                 Brazil      638  Female   
131408        Czech Republic     3635    Male   
423373             Singapore      598  Female   
300667             Lithuania     4482  Female   
96529            Switzerland      516    Male   
79679                 Canada     5210    Male   
317843                Mexico     2141  Female   

                                      ST13Q01 ST14Q01 ST14Q02 ST14Q03 ST14Q04  \
15273                       <ISCED level 3A>      NaN     Yes     NaN     NaN   
313716                       <ISCED level 2>       No      No      No      No   
371898                  <ISCED level 3B, 3C>       No      No      No      No   
55903   She did not complete <ISCED level 1>       No      No      No      No   
131408                  <ISCED level 3B, 3C>       No      No      No      No   
423373                      <ISCED level 3A>      NaN     NaN     NaN     Yes   
300667                      <ISCED level 3A>       No      No     Yes      No   
96529                   <ISCED level 3B, 3C>       No      No      No     Yes   
79679                       <ISCED level 3A>      NaN     Yes     Yes     NaN   
317843                       <ISCED level 1>       No      No      No     NaN   

                      ST17Q01 ST18Q01  ... ST57Q03 ST57Q04 ST57Q05  MMINS  \
15273       <ISCED level 3A>      NaN  ...     2.0     0.0     1.0  240.0   
313716  <ISCED level 3B, 3C>       No  ...     NaN     NaN     NaN    NaN   
371898  <ISCED level 3B, 3C>       No  ...     0.0     0.0     0.0  120.0   
55903        <ISCED level 1>       No  ...     NaN     NaN     NaN    NaN   
131408  <ISCED level 3B, 3C>       No  ...     0.0     0.0     0.0  225.0   
423373       <ISCED level 2>      NaN  ...     NaN     NaN     NaN    NaN   
300667      <ISCED level 3A>       No  ...     0.0     0.0     1.0  225.0   
96529   <ISCED level 3B, 3C>       No  ...     0.0     0.0     0.0  350.0   
79679       <ISCED level 3A>      NaN  ...     0.0     0.0     1.0  300.0   
317843       <ISCED level 1>       No  ...     0.0     0.0     1.0  325.0   

        LMINS  SMINS  Average Math Score  Average Reading Score  \
15273   320.0  200.0           488.86898              470.39440   
313716    NaN    NaN           500.63094              471.76602   
371898  160.0  120.0           515.58656              518.92676   
55903     NaN    NaN           386.51660              409.23250   
131408  180.0   90.0           440.10752              397.58612   
423373    NaN    NaN           479.98908              442.27582   
300667  225.0  225.0           476.09440              490.09328   
96529   350.0  200.0           425.69712              387.24102   
79679   300.0  300.0           395.55228              461.82192   
317843  240.0  300.0           358.47486              418.12880   

        Average Science Score  Average Total Score  
15273               477.75084           479.004740  
313716              463.01752           478.471493  
371898              510.38790           514.967073  
55903               384.12904           393.292713  
131408              439.89184           425.861827  
423373              424.78554           449.016813  
300667              497.51958           487.902420  
96529               408.56024           407.166127  
79679               412.66318           423.345793  
317843              377.22862           384.610760  

[10 rows x 25 columns]

### Make the variable names readable

In the original dataframe, the variable names are all codes and unreadable without the dictionary. Therefore, we will change them to accurately reflect the information within the column.

In [99]:
# Rename columns
pisa_clean.rename({'CNT': 'Country',
                   'STIDSTD': 'Student ID',
                   'ST04Q01': 'Gender', 
                   'ST13Q01': 'Mother - Highest Schooling', 
                   'ST14Q01': 'Mother Qualifications - ISCED level 6',
                   'ST14Q02': 'Mother Qualifications - ISCED level 5A',
                   'ST14Q03': 'Mother Qualifications - ISCED level 5B',
                   'ST14Q04': 'Mother Qualifications - ISCED level 4',
                   'ST17Q01': 'Father - Highest Schooling',
                   'ST18Q01': 'Father Qualifications - ISCED level 6',
                   'ST18Q02': 'Father Qualifications - ISCED level 5A',
                   'ST18Q03': 'Father Qualifications - ISCED level 5B',
                   'ST18Q04': 'Father Qualifications - ISCED level 4',
                   'ST57Q01': 'Out-of-School Study Time - Homework',
                   'ST57Q02': 'Out-of-School Study Time - Guided Homework',
                   'ST57Q03': 'Out-of-School Study Time - Personal Tutor',
                   'ST57Q04': 'Out-of-School Study Time - Commercial Company',
                   'ST57Q05': 'Out-of-School Study Time - With Parent',
                   'MMINS': 'Learning Time - Mathematics',
                   'LMINS': 'Learning Time - Test Language',
                   'SMINS': 'Learning Time - Science'}, axis='columns', inplace=True)

In [100]:
# Check to see if the columns are correctly renamed
list(pisa_clean)

['Country',
 'Student ID',
 'Gender',
 'Mother - Highest Schooling',
 'Mother Qualifications - ISCED level 6',
 'Mother Qualifications - ISCED level 5A',
 'Mother Qualifications - ISCED level 5B',
 'Mother Qualifications - ISCED level 4',
 'Father - Highest Schooling',
 'Father Qualifications - ISCED level 6',
 'Father Qualifications - ISCED level 5A',
 'Father Qualifications - ISCED level 5B',
 'Father Qualifications - ISCED level 4',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning Time - Mathematics',
 'Learning Time - Test Language',
 'Learning Time - Science',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score']

# Null

In [151]:
pisa_clean.replace(["NaN"], np.nan, inplace = True)
pisa_clean.dropna(inplace = True)

In [152]:
pisa_clean.duplicated().sum()

0

### Education
Multiple columns with the single variable. 

http://uis.unesco.org/sites/default/files/documents/international-standard-classification-of-education-isced-2011-en.pdf

In [153]:
father_edu = pisa_clean.copy()

In [154]:
father_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'He did not complete <ISCED level 1>']] = pd.get_dummies(father_edu['Father - Highest Schooling'])

In [155]:
father_edu.rename({'He did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Father Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Father Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [156]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].astype(str)
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].astype(str)
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].astype(str)
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].astype(str)

In [157]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
father_edu['<ISCED level 4>'] = father_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
father_edu['<ISCED level 5>'] = father_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
father_edu['<ISCED level 6>'] = father_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [158]:
father_edu.sample(10)

Country  Student ID  Gender Mother - Highest Schooling  \
286898            Kazakhstan        1847  Female          <ISCED level 3A>    
426940             Singapore        4165    Male          <ISCED level 3A>    
89389                 Canada       14920    Male          <ISCED level 3A>    
408257         Florida (USA)        1103  Female          <ISCED level 3A>    
180259               Finland        4585    Male          <ISCED level 3A>    
80043                 Canada        5574  Female          <ISCED level 3A>    
148214                 Spain        2632  Female          <ISCED level 3A>    
368247                Norway        4338  Female          <ISCED level 3A>    
13912   United Arab Emirates        9170  Female          <ISCED level 3A>    
219730               Hungary        3152    Male      <ISCED level 3B, 3C>    

       Mother Qualifications - ISCED level 6  \
286898                                    No   
426940                                    No   
89389                                     No   
408257                                    No   
180259                                   Yes   
80043                                     No   
148214                                    No   
368247                                    No   
13912                                     No   
219730                                    No   

       Mother Qualifications - ISCED level 5A  \
286898                                     No   
426940                                     No   
89389                                      No   
408257                                     No   
180259                                    Yes   
80043                                      No   
148214                                     No   
368247                                     No   
13912                                      No   
219730                                     No   

       Mother Qualifications - ISCED level 5B  \
286898                                    Yes   
426940                                     No   
89389                                      No   
408257                                     No   
180259                                     No   
80043                                     Yes   
148214                                    Yes   
368247                                     No   
13912                                      No   
219730                                     No   

       Mother Qualifications - ISCED level 4  \
286898                                    No   
426940                                   Yes   
89389                                    Yes   
408257                                    No   
180259                                    No   
80043                                    Yes   
148214                                   Yes   
368247                                   Yes   
13912                                     No   
219730                                    No   

                  Father - Highest Schooling  <ISCED level 6>  ...  \
286898                 <ISCED level 3B, 3C>                No  ...   
426940  He did not complete <ISCED level 1>                No  ...   
89389                      <ISCED level 3A>                No  ...   
408257                     <ISCED level 3A>                No  ...   
180259                     <ISCED level 3A>   <ISCED level 6>  ...   
80043                      <ISCED level 3A>                No  ...   
148214                     <ISCED level 3A>   <ISCED level 6>  ...   
368247                 <ISCED level 3B, 3C>                No  ...   
13912                  <ISCED level 3B, 3C>                No  ...   
219730                 <ISCED level 3B, 3C>                No  ...   

       Learning Time - Science Average Math Score Average Reading Score  \
286898                    90.0          458.49040             512.81056   
426940                   400.0          656.41846             597.91198   
89389            

In [159]:
father_edu = pd.melt(father_edu, 
                     id_vars=['Country','Student ID', 'Gender',
                            'Mother - Highest Schooling', 
                            'Mother Qualifications - ISCED level 6', 
                            'Mother Qualifications - ISCED level 5A',
                            'Mother Qualifications - ISCED level 5B',
                            'Mother Qualifications - ISCED level 4',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning Time - Mathematics',
                            'Learning Time - Test Language',
                            'Learning Time - Science',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score'], var_name='Completed_edu - Father', value_name='Education - Father')


In [160]:
father_edu.sample(10)

Country  Student ID  Gender  \
285905   Hong Kong-China        1595    Male   
998224         Lithuania        2311  Female   
828311       Switzerland        3969    Male   
699797          Bulgaria        1163  Female   
1140052        Singapore        3875  Female   
766761        Kazakhstan        1947    Male   
755192             Italy        2047    Male   
798188   Slovak Republic        4224  Female   
321382              Peru        1357    Male   
198180            Latvia        1516    Male   

                    Mother - Highest Schooling  \
285905                       <ISCED level 3A>    
998224                       <ISCED level 3A>    
828311                        <ISCED level 2>    
699797                   <ISCED level 3B, 3C>    
1140052                      <ISCED level 3A>    
766761                       <ISCED level 3A>    
755192                        <ISCED level 2>    
798188                       <ISCED level 3A>    
321382   She did not complete <ISCED level 1>    
198180                       <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
285905                                     No   
998224                                     No   
828311                                     No   
699797                                     No   
1140052                                    No   
766761                                     No   
755192                                     No   
798188                                     No   
321382                                     No   
198180                                     No   

        Mother Qualifications - ISCED level 5A  \
285905                                     Yes   
998224                                      No   
828311                                      No   
699797                                      No   
1140052                                     No   
766761                                      No   
755192                                      No   
798188                                      No   
321382                                      No   
198180                                     Yes   

        Mother Qualifications - ISCED level 5B  \
285905                                      No   
998224                                      No   
828311                                      No   
699797                                     Yes   
1140052                                     No   
766761                                     Yes   
755192                                      No   
798188                                      No   
321382                                      No   
198180                                      No   

        Mother Qualifications - ISCED level 4  \
285905                                     No   
998224                                     No   
828311                                     No   
699797                                     No   
1140052                                    No   
766761                                     No   
755192                                     No   
798188                                     No   
321382                                     No   
198180                                     No   

         Out-of-School Study Time - Homework  \
285905                                   4.0   
998224                                   2.0   
828311                                   1.0   
699797                                   5.0   
1140052                                  4.0   
766761                                   1.0   
755192                                   3.0   
798188                                   3.0   
321382                                   4.0   
198180                                   4.0   

         Out-of-School Study Time - Guided Homework  ...  \
285905                                          0.0  ...   
998224                                          1.0  ...   
828311                                        

In [161]:
father_edu = father_edu[(father_edu['Education - Father'] != '0')] 
father_edu = father_edu[(father_edu['Education - Father'] != 'No')]

In [162]:
father_edu.sample(70)

Country  Student ID  Gender  \
889297                   Malaysia         927  Female   
259266                      Chile        6408  Female   
660993                Netherlands         904    Male   
566665                  Singapore        3996    Male   
60138                     Hungary         226  Female   
929693                   Bulgaria        3241  Female   
489264                 Costa Rica        2738    Male   
60640                     Hungary        1412    Male   
895985                     Poland         942    Male   
844006                      Spain       11729    Male   
36135                     Denmark        4348    Male   
944484                Switzerland        7733  Female   
887194                Macao-China         594    Male   
36623                     Denmark        5565  Female   
88922                      Norway         425    Male   
997105                 Kazakhstan        5247    Male   
501742                      Spain       17151    Male   
727438                      Spain        6020  Female   
618539                      Spain       23612    Male   
6567                    Australia       12643    Male   
901072             China-Shanghai         235    Male   
263574                    Germany        4114    Male   
398252             United Kingdom        9662    Male   
571755   United States of America         406    Male   
96242                    Portugal        3186  Female   
566024                  Singapore        2496    Male   
906506                    Romania        2039  Female   
18036                      Canada        4586  Female   
917488   United States of America        4936  Female   
805241       United Arab Emirates        7206    Male   
...                           ...         ...     ...   
945354                Switzerland        9926    Male   
52967              United Kingdom        6409  Female   
378275                    Germany        4164    Male   
1008489               New Zealand        2084    Male   
87086                    Malaysia        4251  Female   
891113                Netherlands        3029  Female   
929434                   Bulgaria        2197  Female   
856753             United Kingdom        8986    Male   
998044                  Lithuania        1879    Male   
42667                       Spain       16209    Male   
30548                  Costa Rica        2915  Female   
94407                      Poland        2986    Male   
984269                      Italy        1253  Female   
575491       United Arab Emirates        5389    Male   
310514                  Lithuania        3459    Male   
929915                   Bulgaria        4060    Male   
35490                     Denmark        2741    Male   
550810                       Peru        1735    Male   
449364         Russian Federation        1804  Female   
812172                    Belgium        1933  Female   
474450                     Brazil       15297    Male   
910254                  Singapore        2888    Male   
300571                      Italy       12935    Male   
775828                Netherlands        1349    Male   
68024                       Italy        4596    Male   
26                        Albania         166    Male   
59552                     Croatia        3802  Female   
97353                       Qatar        4186  Female   
26018                 Switzerland        5233  Female   
875045                      Italy       15605  Female   

                    Mother - Highest Schooling  \
889297                       <ISCED level 3A>    
259266                       <ISCED level 3A>    
660993                       <ISCED level 3A>    
566665                        <ISCED level 1>    
60138                        <ISCED level 3A>    
929693                   <ISCED level 3B, 3C>    
489264                       <ISCED level 3A>    
60640                        <ISCED level 3A>    
895985                       <ISCED level

In [163]:
father_edu = father_edu[(father_edu['Completed_edu - Father'] != 'Father - Highest Schooling')] 

In [164]:
father_edu['Education - Father'].value_counts()

<ISCED level 3>    84496
<ISCED level 5>    49085
<ISCED level 4>    26411
<ISCED level 2>    21955
<ISCED level 6>     6715
<ISCED level 1>     5947
<ISCED level 0>     2290
Name: Education - Father, dtype: int64

In [165]:
father_edu.drop(columns=['Completed_edu - Father'], axis=1, inplace=True)

In [166]:
father_edu.sample(10)

Country  Student ID  Gender Mother - Highest Schooling  \
561917    Romania         492    Male          <ISCED level 3A>    
965256    Finland         570    Male           <ISCED level 2>    
223261  Singapore        5509    Male          <ISCED level 3A>    
863429    Hungary        1364    Male      <ISCED level 3B, 3C>    
379483    Denmark        2590    Male          <ISCED level 3A>    
447740    Romania        1981  Female          <ISCED level 3A>    
743880     Greece        4420  Female           <ISCED level 2>    
729785      Spain       13196  Female      <ISCED level 3B, 3C>    
498794      Spain        8190    Male      <ISCED level 3B, 3C>    
454434   Slovenia         460  Female           <ISCED level 2>    

       Mother Qualifications - ISCED level 6  \
561917                                    No   
965256                                    No   
223261                                   Yes   
863429                                    No   
379483                                    No   
447740                                    No   
743880                                    No   
729785                                    No   
498794                                    No   
454434                                    No   

       Mother Qualifications - ISCED level 5A  \
561917                                     No   
965256                                     No   
223261                                    Yes   
863429                                     No   
379483                                     No   
447740                                    Yes   
743880                                     No   
729785                                     No   
498794                                     No   
454434                                     No   

       Mother Qualifications - ISCED level 5B  \
561917                                     No   
965256                                     No   
223261                                    Yes   
863429                                    Yes   
379483                                    Yes   
447740                                    Yes   
743880                                     No   
729785                                    Yes   
498794                                    Yes   
454434                                     No   

       Mother Qualifications - ISCED level 4  \
561917                                    No   
965256                                    No   
223261                                   Yes   
863429                                   Yes   
379483                                    No   
447740                                    No   
743880                                    No   
729785                                    No   
498794                                   Yes   
454434                                    No   

        Out-of-School Study Time - Homework  \
561917                                  3.0   
965256                                  0.0   
223261                                  1.0   
863429                                 20.0   
379483                                  4.0   
447740                                  8.0   
743880                                  3.0   
729785                                 10.0   
498794                                  3.0   
454434                                  0.0   

        Out-of-School Study Time - Guided Homework  ...  \
561917                                         1.0  ...   
965256                                         0.0  ...   
223261                                         1.0  ...   
863429                                        15.0  ...   
379483                                         1.0  ...   
447740                                         0.0  ...   
743880                                         0.0  ...   
729785                                         4.0  ...   
498794                                         2.0  ...   
454434         

In [167]:
father_edu.duplicated().sum()

6947

In [168]:
# https://stackoverflow.com/questions/33042777/removing-duplicates-from-pandas-dataframe-with-condition-for-retaining-original
father_edu['Education - Father'] = father_edu['Education - Father'].astype('category')

father_edu['Education - Father'] = father_edu['Education - Father'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
father_edu.sort_values(['Education - Father'], inplace=True)
# father_edu_clean = father_edu.drop_duplicates(subset='Student ID', keep='last')

In [169]:
father_edu_clean.duplicated().sum()

0

In [170]:
father_edu.shape

(196899, 21)

In [171]:
father_edu_clean.sample(10)

Country  Student ID  Gender             Mother - Highest Schooling  \
1312598  Mexico       25916  Female                       <ISCED level 1>    
1234766   Italy       23197  Female                      <ISCED level 3A>    
1315741  Mexico       29059    Male  She did not complete <ISCED level 1>    
1320155  Mexico       33473    Male                       <ISCED level 1>    
1299705  Mexico       13023  Female  She did not complete <ISCED level 1>    
1312316  Mexico       25634    Male                       <ISCED level 2>    
2269247  Mexico       11585  Female                       <ISCED level 2>    
652077    Spain       21005    Male                       <ISCED level 2>    
650365    Spain       19293  Female                  <ISCED level 3B, 3C>    
649264    Spain       18192    Male                                    NaN   

        Mother Qualifications - ISCED level 6  \
1312598                                    No   
1234766                                   NaN   
1315741                                    No   
1320155                                   NaN   
1299705                                    No   
1312316                                    No   
2269247                                    No   
652077                                     No   
650365                                    NaN   
649264                                    NaN   

        Mother Qualifications - ISCED level 5A  \
1312598                                     No   
1234766                                    NaN   
1315741                                    NaN   
1320155                                    NaN   
1299705                                    NaN   
1312316                                     No   
2269247                                     No   
652077                                      No   
650365                                     NaN   
649264                                     NaN   

        Mother Qualifications - ISCED level 5B  \
1312598                                     No   
1234766                                    NaN   
1315741                                    NaN   
1320155                                    NaN   
1299705                                    NaN   
1312316                                     No   
2269247                                     No   
652077                                      No   
650365                                     NaN   
649264                                     NaN   

        Mother Qualifications - ISCED level 4  \
1312598                                   NaN   
1234766                                   NaN   
1315741                                   NaN   
1320155                                   NaN   
1299705                                   NaN   
1312316                                   NaN   
2269247                                   NaN   
652077                                     No   
650365                                    Yes   
649264                                    NaN   

         Out-of-School Study Time - Homework  \
1312598                                  4.0   
1234766                                  NaN   
1315741                                  4.0   
1320155                                  NaN   
1299705                                  NaN   
1312316                                  NaN   
2269247                                  4.0   
652077                                   7.0   
650365                                   NaN   
649264                                   3.0   

         Out-of-School Study Time - Guided Homework  ...  \
1312598                                         2.0  ...   
1234766                                         NaN  ...   
1315741                                         NaN  ...   
1320155                                         NaN  ...   
1299705                                         NaN  ...   
1312316                                         NaN  ...   
2269247           

### Mother Edu

In [172]:
mother_edu = father_edu.copy()

In [173]:
mother_edu.head()

Country  Student ID  Gender  \
1146860  United States of America        4922  Female   
1062200                Costa Rica          46    Male   
1062201                Costa Rica          48  Female   
1062215                Costa Rica         100  Female   
1062219                Costa Rica         126    Male   

                    Mother - Highest Schooling  \
1146860  She did not complete <ISCED level 1>    
1062200  She did not complete <ISCED level 1>    
1062201                       <ISCED level 1>    
1062215                       <ISCED level 1>    
1062219                      <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
1146860                                    No   
1062200                                    No   
1062201                                    No   
1062215                                    No   
1062219                                    No   

        Mother Qualifications - ISCED level 5A  \
1146860                                     No   
1062200                                     No   
1062201                                     No   
1062215                                     No   
1062219                                     No   

        Mother Qualifications - ISCED level 5B  \
1146860                                     No   
1062200                                     No   
1062201                                     No   
1062215                                     No   
1062219                                     No   

        Mother Qualifications - ISCED level 4  \
1146860                                    No   
1062200                                    No   
1062201                                    No   
1062215                                    No   
1062219                                    No   

         Out-of-School Study Time - Homework  \
1146860                                 10.0   
1062200                                 10.0   
1062201                                  4.0   
1062215                                  5.0   
1062219                                  1.0   

         Out-of-School Study Time - Guided Homework  ...  \
1146860                                         0.0  ...   
1062200                                         1.0  ...   
1062201                                         2.0  ...   
1062215                                         0.0  ...   
1062219                                         0.0  ...   

         Out-of-School Study Time - Commercial Company  \
1146860                                            0.0   
1062200                                            2.0   
1062201                                            0.0   
1062215                                            0.0   
1062219                                            0.0   

         Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
1146860                                     0.0                        300.0   
1062200                                     1.0                        320.0   
1062201                                     1.0                        160.0   
1062215                                     0.0                        120.0   
1062219                                     0.0                        200.0   

         Learning Time - Test Language  Learning Time - Science  \
1146860                          300.0                    300.0   
1062200                          240.0                    240.0   
1062201                          240.0                    240.0   
1062215                          120.0                    240.0   
1062219                          200.0                    360.0   

         Average Math Score  Average Reading Score  Average Science Score  \
1146860           596.82972              578.02342              547.40770   
1062200           274.27170              302.31502              343.84560   
1062201           343.98660              357.76122              351.30548   

In [174]:
mother_edu['Mother - Highest Schooling'].value_counts()

<ISCED level 3A>                         120377
<ISCED level 3B, 3C>                      39451
<ISCED level 2>                           28132
<ISCED level 1>                            6189
She did not complete <ISCED level 1>       2750
Name: Mother - Highest Schooling, dtype: int64

In [175]:
pisa_clean['Mother - Highest Schooling'].value_counts()

<ISCED level 3A>                         63660
<ISCED level 3B, 3C>                     23596
<ISCED level 2>                          19860
<ISCED level 1>                           5189
She did not complete <ISCED level 1>      2383
Name: Mother - Highest Schooling, dtype: int64

In [176]:
mother_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'She did not complete <ISCED level 1>']] = pd.get_dummies(mother_edu['Mother - Highest Schooling'])

In [177]:
mother_edu.rename({'She did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Mother Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Mother Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [178]:
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].astype(str)
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].astype(str)
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].astype(str)
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].astype(str)

In [179]:
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
mother_edu['<ISCED level 4>'] = mother_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
mother_edu['<ISCED level 5>'] = mother_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
mother_edu['<ISCED level 6>'] = mother_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [180]:
mother_edu.sample(10)

Country  Student ID  Gender  \
694415                 Australia       11538    Male   
837610                   Denmark         976    Male   
687660  United States of America        3744  Female   
983151                    Israel        2698  Female   
752466                   Iceland        1693  Female   
972143            United Kingdom       10928    Male   
866859                   Iceland         822  Female   
361002                    Canada        1435    Male   
879235                     Italy       26603  Female   
574991      United Arab Emirates        3086    Male   

       Mother - Highest Schooling <ISCED level 6>  <ISCED level 5>  \
694415          <ISCED level 3A>               No  <ISCED level 5>   
837610          <ISCED level 3A>               No               No   
687660          <ISCED level 3A>               No               No   
983151          <ISCED level 3A>               No               No   
752466           <ISCED level 2>               No               No   
972143      <ISCED level 3B, 3C>               No               No   
866859          <ISCED level 3A>               No               No   
361002          <ISCED level 3A>               No  <ISCED level 5>   
879235          <ISCED level 3A>               No               No   
574991          <ISCED level 3A>               No               No   

        <ISCED level 5>  <ISCED level 4>  Out-of-School Study Time - Homework  \
694415               No               No                                  7.0   
837610  <ISCED level 5>               No                                  2.0   
687660               No               No                                  3.0   
983151  <ISCED level 5>               No                                  6.0   
752466               No               No                                  3.0   
972143  <ISCED level 5>  <ISCED level 4>                                  7.0   
866859               No               No                                  2.0   
361002  <ISCED level 5>               No                                  1.0   
879235               No               No                                 15.0   
574991               No               No                                  4.0   

        Out-of-School Study Time - Guided Homework  ...  Average Math Score  \
694415                                         0.0  ...           490.58262   
837610                                         0.0  ...           580.54994   
687660                                         0.0  ...           521.97382   
983151                                         0.0  ...           572.99424   
752466                                         0.0  ...           471.88814   
972143                                         0.0  ...           532.64526   
866859                                         1.0  ...           381.92086   
361002                                         0.0  ...           631.41456   
879235                                         2.0  ...           478.58700   
574991                                         4.0  ...           349.04968   

        Average Reading Score  Average Science Score  Average Total Score  \
694415              498.15004              538.54906           509.093907   
837610              575.21694              565.59120           573.786027   
687660              576.11708              563.53972           553.876873   
983151              610.82842              535.93808           573.253580   
752466              561.74006              509.54866           514.392287   
972143              553.08328              583.30848           556.345673   
866859              390.24850              366.22528           379.464880   
361002              570.40528              539.94778           580.589207   
879235              563.96412              526.05372           522.868280   
574991              328.77922              346.45656           341.428487   

        Education - Father 

In [181]:
mother_edu = pd.melt(mother_edu, 
                     id_vars=['Country','Student ID', 'Gender',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning Time - Mathematics',
                            'Learning Time - Test Language',
                            'Learning Time - Science',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score',
                             'Education - Father'], var_name='Completed_edu - Mother', value_name='Education - Mother')


In [182]:
mother_edu.sample(10)

Country  Student ID  Gender  \
961343   United Arab Emirates         714  Female   
1257000               Austria        1046    Male   
493938                 Norway        1865    Male   
843446                  Qatar        5050  Female   
918137                Belgium        7827  Female   
53839             Netherlands        1471  Female   
69529                 Estonia        1001  Female   
1382650             Singapore        4575    Male   
161768                  Spain       20625    Male   
81391             Switzerland        6920  Female   

         Out-of-School Study Time - Homework  \
961343                                   7.0   
1257000                                  6.0   
493938                                   5.0   
843446                                   6.0   
918137                                  10.0   
53839                                   10.0   
69529                                    3.0   
1382650                                  5.0   
161768                                  10.0   
81391                                   10.0   

         Out-of-School Study Time - Guided Homework  \
961343                                          2.0   
1257000                                         4.0   
493938                                          1.0   
843446                                          0.0   
918137                                          0.0   
53839                                           0.0   
69529                                           1.0   
1382650                                         4.0   
161768                                          0.0   
81391                                           0.0   

         Out-of-School Study Time - Personal Tutor  \
961343                                         6.0   
1257000                                        2.0   
493938                                         0.0   
843446                                         1.0   
918137                                         0.0   
53839                                          0.0   
69529                                          0.0   
1382650                                        0.0   
161768                                         0.0   
81391                                          0.0   

         Out-of-School Study Time - Commercial Company  \
961343                                             6.0   
1257000                                            2.0   
493938                                             0.0   
843446                                             0.0   
918137                                             0.0   
53839                                              0.0   
69529                                              0.0   
1382650                                            0.0   
161768                                             0.0   
81391                                              0.0   

         Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
961343                                      0.0                        360.0   
1257000                                     0.0                        150.0   
493938                                      1.0                        120.0   
843446                                      1.0                        275.0   
918137                                      3.0                        250.0   
53839                                       0.0                        120.0   
69529                                       1.0                        225.0   
1382650                                     0.0                        385.0   
161768                                      0.0                        220.0   
81391                                       0.0                        250.0   

         Learning Time - Test Language  Learning Time - Science  \
961343                           180.0                      0.0   
1257000                          150.0                

In [183]:
mother_edu = mother_edu[(mother_edu['Education - Mother'] != '0')] 
mother_edu = mother_edu[(mother_edu['Education - Mother'] != 'No')]

In [184]:
mother_edu.sample(70)

Country  Student ID  Gender  \
1514072                    Israel        2212  Female   
1447720                   Estonia         737    Male   
186410                      Spain        8548  Female   
1038654               Macao-China        5331  Female   
177311                  Australia        2021    Male   
183368                 Kazakhstan         797    Male   
156101                Switzerland        4381  Female   
8501                    Australia       12223  Female   
57635                      Poland         966  Female   
907032                 Luxembourg        1569    Male   
1531601                   Romania        4411    Male   
49744                     Romania        4328    Male   
519688          Connecticut (USA)        1347    Male   
135190            Hong Kong-China        3131  Female   
1759341                 Lithuania        3105    Male   
122142                   Slovenia        3562  Female   
148331            Slovak Republic        4329    Male   
646852                      Qatar        6520  Female   
717569                     Canada       17251    Male   
1538769                   Estonia        3217    Male   
96198                     Denmark        1229  Female   
4419                    Singapore        3570  Female   
787550                   Malaysia        4879    Male   
106006                      Italy        6469    Male   
691437                     Norway         353    Male   
170866                    Denmark        3648  Female   
1198466                  Malaysia        1880  Female   
138377                    Denmark        1994  Female   
72626                      Brazil       14869  Female   
1426                       Poland        1137  Female   
...                           ...         ...     ...   
113360                   Slovenia        4325    Male   
1424594            China-Shanghai        2408    Male   
896365                  Lithuania        1383  Female   
1669442                   Austria        1443  Female   
1448902            United Kingdom       10305    Male   
568605                    Romania        1239  Female   
1412236                 Lithuania        3093  Female   
147170                      Italy       15675    Male   
564367                    Denmark        6043    Male   
194061                      Chile        4831    Male   
353851                     Canada       18554    Male   
178548                      Italy       18497  Female   
1423929                   Romania        1552    Male   
23208                       Italy       28449    Male   
836689            Slovak Republic         456    Male   
15373             Slovak Republic        2129  Female   
190512                   Bulgaria        1032    Male   
1540113                     Spain       17640  Female   
1712266                   Iceland        2997  Female   
1676691               New Zealand        1294  Female   
1457132                    Canada       21215    Male   
841010                     Latvia        2231  Female   
793761                  Australia       13104  Female   
1207389                     Spain       10114    Male   
982038                      Spain        8511    Male   
1422591       Massachusetts (USA)        1238  Female   
108330                 Luxembourg        1731  Female   
1716581                Luxembourg        2936    Male   
1733669      United Arab Emirates        3169    Male   
157981   United States of America        3265    Male   

         Out-of-School Study Time - Homework  \
1514072                                  8.0   
1447720                                  8.0   
186410                                   3.0   
1038654                                  1.0   
177311                                  10.0   
183368                                   9.0   
156101                                   2.0   
8501                                     1.0   
57635                                    3.0   
907032     

In [185]:
mother_edu = mother_edu[(mother_edu['Completed_edu - Mother'] != 'Mother - Highest Schooling')] 

In [186]:
mother_edu['Education - Mother'].value_counts()

<ISCED level 3>    159828
<ISCED level 5>    115513
<ISCED level 4>     55348
<ISCED level 2>     28132
<ISCED level 6>     14532
<ISCED level 1>      6189
<ISCED level 0>      2750
Name: Education - Mother, dtype: int64

In [187]:
mother_edu.drop(columns=['Completed_edu - Mother'], axis=1, inplace=True)

In [188]:
mother_edu.sample(10)

Country  Student ID  Gender  \
521779               Canada        7271  Female   
433266            Australia        7449  Female   
1561185          Kazakhstan        5617    Male   
816663                Spain       21684  Female   
1426950     Slovak Republic        2120  Female   
536643          New Zealand         846    Male   
445517               Norway         103    Male   
1001581            Malaysia        1360  Female   
1552936  Russian Federation        1803  Female   
895635                Italy        8670  Female   

         Out-of-School Study Time - Homework  \
521779                                   2.0   
433266                                  10.0   
1561185                                 12.0   
816663                                   8.0   
1426950                                  3.0   
536643                                   5.0   
445517                                   4.0   
1001581                                  3.0   
1552936                                  4.0   
895635                                  21.0   

         Out-of-School Study Time - Guided Homework  \
521779                                          0.0   
433266                                          0.0   
1561185                                         0.0   
816663                                          2.0   
1426950                                         0.0   
536643                                          4.0   
445517                                          0.0   
1001581                                         2.0   
1552936                                         3.0   
895635                                          2.0   

         Out-of-School Study Time - Personal Tutor  \
521779                                         0.0   
433266                                         0.0   
1561185                                        0.0   
816663                                         4.0   
1426950                                        0.0   
536643                                         0.0   
445517                                         0.0   
1001581                                        1.0   
1552936                                        2.0   
895635                                         2.0   

         Out-of-School Study Time - Commercial Company  \
521779                                             0.0   
433266                                             0.0   
1561185                                            0.0   
816663                                             4.0   
1426950                                            0.0   
536643                                             0.0   
445517                                             0.0   
1001581                                            3.0   
1552936                                            0.0   
895635                                             0.0   

         Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
521779                                      0.0                        375.0   
433266                                      0.0                        232.0   
1561185                                     0.0                        225.0   
816663                                      0.0                        165.0   
1426950                                     0.0                        225.0   
536643                                      4.0                        200.0   
445517                                      0.0                        180.0   
1001581                                     4.0                        320.0   
1552936                                     0.0                        135.0   
895635                                      0.0                        300.0   

         Learning Time - Test Language  Learning Time - Science  \
521779                             0.0                    375.0   
433266                           232.0                    232.0   
1561185

In [189]:
mother_edu['Student ID'].duplicated().sum()

358707

In [190]:
# mother_edu.replace(["NaN"], np.nan, inplace = True)
# mother_edu.dropna(inplace = True)

In [191]:
mother_edu['Education - Mother'] = mother_edu['Education - Mother'].astype('category')

mother_edu['Education - Mother'] = mother_edu['Education - Mother'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
mother_edu.sort_values(['Education - Mother'], inplace=True)
mother_edu_clean = mother_edu.drop_duplicates(subset='Student ID', keep='last')

In [192]:
mother_edu['Student ID'].duplicated().sum()

358707

In [193]:
mother_edu.shape

(382292, 17)

In [194]:
mother_edu_clean.shape

(23585, 17)

In [195]:
mother_edu_clean.duplicated().sum()

0

In [196]:
mother_edu_clean.sample(10)

Country  Student ID  Gender  Out-of-School Study Time - Homework  \
967364    Italy       25736    Male                                  4.0   
269995   Brazil        5275  Female                                  2.0   
1410099   Italy       23364  Female                                 12.0   
905579    Italy       29462    Male                                 24.0   
1203354   Italy       24082  Female                                  5.0   
1711213   Italy        7704    Male                                 12.0   
472545   Canada       20961    Male                                  3.0   
673216   Canada        8918  Female                                  1.0   
417537    Italy       12040    Male                                  0.0   
472842   Canada       20484    Male                                  2.0   

         Out-of-School Study Time - Guided Homework  \
967364                                          4.0   
269995                                          2.0   
1410099                                         0.0   
905579                                          0.0   
1203354                                         0.0   
1711213                                         4.0   
472545                                          1.0   
673216                                          0.0   
417537                                          0.0   
472842                                          1.0   

         Out-of-School Study Time - Personal Tutor  \
967364                                         0.0   
269995                                         2.0   
1410099                                        0.0   
905579                                         0.0   
1203354                                        0.0   
1711213                                        0.0   
472545                                         0.0   
673216                                         0.0   
417537                                         0.0   
472842                                         2.0   

         Out-of-School Study Time - Commercial Company  \
967364                                             0.0   
269995                                             3.0   
1410099                                            0.0   
905579                                             3.0   
1203354                                            0.0   
1711213                                            0.0   
472545                                             0.0   
673216                                             0.0   
417537                                             0.0   
472842                                             1.0   

         Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
967364                                      0.0                        250.0   
269995                                      0.0                        200.0   
1410099                                     0.0                        250.0   
905579                                      0.0                        300.0   
1203354                                     1.0                        180.0   
1711213                                     6.0                        300.0   
472545                                      0.0                        275.0   
673216                                      0.0                        225.0   
417537                                      0.0                        150.0   
472842                                      2.0                         75.0   

         Learning Time - Test Language  Learning Time - Science  \
967364                           250.0                    200.0   
269995                           200.0                    150.0   
1410099                          200.0                    100.0   
905579                           240.0                    120.0   
1203354                          360.0                    180.0   
1711213                          420.0          

In [197]:
pisa_clean = mother_edu_clean.copy()

In [198]:
pisa_clean.shape

(23585, 17)

In [199]:
pisa_clean.duplicated().sum()

0

In [200]:
pisa_clean = mother_edu.copy()

In [201]:
pisa_clean.shape

(382292, 17)

### 4. Missing values in relevant columns 
('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')

Due to the amount of rows in this dataframe, we can comfortably remove the rows with missing values.

In [202]:

# pisa_clean.replace(["NaN"], np.nan, inplace = True)
# pisa_clean.dropna(inplace = True)

In [203]:
# Check to see if null values correctly dropped
pisa_clean

Country  Student ID  Gender  \
1968985           Netherlands        1647    Male   
1774983           Macao-China        2513  Female   
1774972           Macao-China        2954    Male   
1774965           Macao-China        3032  Female   
1774961           Macao-China        3094    Male   
1774944           Macao-China        2717    Male   
1774939           Macao-China        2798    Male   
1774932              Malaysia        1261    Male   
1774923           Macao-China        5249  Female   
1774912              Malaysia        1227  Female   
1774910              Malaysia         879  Female   
1774905           Macao-China        5314  Female   
1774900              Malaysia         126    Male   
1774898              Malaysia         383  Female   
1774892              Malaysia         807  Female   
1774890              Malaysia        2196    Male   
1774880              Malaysia        4236    Male   
1774879              Malaysia        4250  Female   
1774990           Macao-China        2290    Male   
1775026           Macao-China        2368  Female   
1775031           Macao-China        4049  Female   
1775051           Macao-China        3844  Female   
1775322       Hong Kong-China        1977    Male   
1775321       Hong Kong-China        1994    Male   
1775319       Hong Kong-China        2001    Male   
1775200       Hong Kong-China        4187  Female   
1775185       Hong Kong-China        3842  Female   
1775182       Hong Kong-China        3855    Male   
1775166       Hong Kong-China        4633    Male   
1775147               Hungary        1200  Female   
...                       ...         ...     ...   
329541   United Arab Emirates        6439    Male   
329542   United Arab Emirates        6450    Male   
329544   United Arab Emirates        6835    Male   
329549   United Arab Emirates        7218  Female   
329407   United Arab Emirates        7592    Male   
329400   United Arab Emirates        7793    Male   
329387   United Arab Emirates        8054    Male   
329377   United Arab Emirates        7895    Male   
329201   United Arab Emirates        5712    Male   
329231   United Arab Emirates        5281    Male   
329232   United Arab Emirates        5283    Male   
329237   United Arab Emirates        5366    Male   
329241   United Arab Emirates        5498    Male   
329253   United Arab Emirates        4356    Male   
329263   United Arab Emirates        3277    Male   
329276   United Arab Emirates        3424  Female   
329278   United Arab Emirates        3435  Female   
338206             Luxembourg        3269  Female   
329282   United Arab Emirates        3167    Male   
329302   United Arab Emirates        3062  Female   
329304   United Arab Emirates        3109    Male   
329314   United Arab Emirates        4012    Male   
329322   United Arab Emirates        4112  Female   
329340   United Arab Emirates        3984    Male   
329343   United Arab Emirates        3483  Female   
329353   United Arab Emirates        3655    Male   
329364   United Arab Emirates        3789  Female   
329374   United Arab Emirates        7835  Female   
329289   United Arab Emirates        2893  Female   
196993                Denmark         134    Male   

         Out-of-School Study Time - Homework  \
1968985                                  2.0   
1774983                                  5.0   
1774972                                 10.0   
1774965                                 12.0   
1774961                                 12.0   
1774944                                  1.0   
1774939                                  1.0   
1774932                                 14.0   
1774923                                 12.0   
1774912                                  3.0   
1774910                                  5.0   
1774905                                 23.0   
1774900                                  2.0   
1774898                                 10.0   
1774892        

In [204]:
# Confirm with smaller shape of dataframe
pisa_clean.shape

(382292, 17)

### 5. Create a total for Out-of-School Study Time and Learning time (minutes per week)



In [205]:
# Create total of Out-of-School Study Time
pisa_clean['Out-of-School Study Time - Total'] = pisa_clean['Out-of-School Study Time - Homework'] + pisa_clean['Out-of-School Study Time - Guided Homework'] + pisa_clean['Out-of-School Study Time - Personal Tutor'] + pisa_clean['Out-of-School Study Time - Commercial Company'] + pisa_clean['Out-of-School Study Time - With Parent']

# Create total of Learning time (minutes per week)
pisa_clean['Learning Time - Total'] = pisa_clean['Learning Time - Mathematics'] + pisa_clean['Learning Time - Test Language'] + pisa_clean['Learning Time - Science'] 
                                                                                                                 

In [206]:
list(pisa_clean)

['Country',
 'Student ID',
 'Gender',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning Time - Mathematics',
 'Learning Time - Test Language',
 'Learning Time - Science',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score',
 'Education - Father',
 'Education - Mother',
 'Out-of-School Study Time - Total',
 'Learning Time - Total']

<a id='assessing2'></a>
### Final Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [207]:
pisa_clean

Country  Student ID  Gender  \
1968985           Netherlands        1647    Male   
1774983           Macao-China        2513  Female   
1774972           Macao-China        2954    Male   
1774965           Macao-China        3032  Female   
1774961           Macao-China        3094    Male   
1774944           Macao-China        2717    Male   
1774939           Macao-China        2798    Male   
1774932              Malaysia        1261    Male   
1774923           Macao-China        5249  Female   
1774912              Malaysia        1227  Female   
1774910              Malaysia         879  Female   
1774905           Macao-China        5314  Female   
1774900              Malaysia         126    Male   
1774898              Malaysia         383  Female   
1774892              Malaysia         807  Female   
1774890              Malaysia        2196    Male   
1774880              Malaysia        4236    Male   
1774879              Malaysia        4250  Female   
1774990           Macao-China        2290    Male   
1775026           Macao-China        2368  Female   
1775031           Macao-China        4049  Female   
1775051           Macao-China        3844  Female   
1775322       Hong Kong-China        1977    Male   
1775321       Hong Kong-China        1994    Male   
1775319       Hong Kong-China        2001    Male   
1775200       Hong Kong-China        4187  Female   
1775185       Hong Kong-China        3842  Female   
1775182       Hong Kong-China        3855    Male   
1775166       Hong Kong-China        4633    Male   
1775147               Hungary        1200  Female   
...                       ...         ...     ...   
329541   United Arab Emirates        6439    Male   
329542   United Arab Emirates        6450    Male   
329544   United Arab Emirates        6835    Male   
329549   United Arab Emirates        7218  Female   
329407   United Arab Emirates        7592    Male   
329400   United Arab Emirates        7793    Male   
329387   United Arab Emirates        8054    Male   
329377   United Arab Emirates        7895    Male   
329201   United Arab Emirates        5712    Male   
329231   United Arab Emirates        5281    Male   
329232   United Arab Emirates        5283    Male   
329237   United Arab Emirates        5366    Male   
329241   United Arab Emirates        5498    Male   
329253   United Arab Emirates        4356    Male   
329263   United Arab Emirates        3277    Male   
329276   United Arab Emirates        3424  Female   
329278   United Arab Emirates        3435  Female   
338206             Luxembourg        3269  Female   
329282   United Arab Emirates        3167    Male   
329302   United Arab Emirates        3062  Female   
329304   United Arab Emirates        3109    Male   
329314   United Arab Emirates        4012    Male   
329322   United Arab Emirates        4112  Female   
329340   United Arab Emirates        3984    Male   
329343   United Arab Emirates        3483  Female   
329353   United Arab Emirates        3655    Male   
329364   United Arab Emirates        3789  Female   
329374   United Arab Emirates        7835  Female   
329289   United Arab Emirates        2893  Female   
196993                Denmark         134    Male   

         Out-of-School Study Time - Homework  \
1968985                                  2.0   
1774983                                  5.0   
1774972                                 10.0   
1774965                                 12.0   
1774961                                 12.0   
1774944                                  1.0   
1774939                                  1.0   
1774932                                 14.0   
1774923                                 12.0   
1774912                                  3.0   
1774910                                  5.0   
1774905                                 23.0   
1774900                                  2.0   
1774898                                 10.0   
1774892        

In [208]:
pisa_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 382292 entries, 1968985 to 196993
Data columns (total 19 columns):
Country                                          382292 non-null object
Student ID                                       382292 non-null int64
Gender                                           382292 non-null object
Out-of-School Study Time - Homework              382292 non-null float64
Out-of-School Study Time - Guided Homework       382292 non-null float64
Out-of-School Study Time - Personal Tutor        382292 non-null float64
Out-of-School Study Time - Commercial Company    382292 non-null float64
Out-of-School Study Time - With Parent           382292 non-null float64
Learning Time - Mathematics                      382292 non-null float64
Learning Time - Test Language                    382292 non-null float64
Learning Time - Science                          382292 non-null float64
Average Math Score                               382292 non-null float64
Average Rea

In [209]:
pisa_clean.shape

(382292, 19)

In [210]:
pisa_clean.drop_duplicates(keep='last', inplace=True)

In [211]:
pisa_clean.duplicated().sum()

0

In [212]:
pisa_clean['Student ID'].drop_duplicates(keep='last', inplace=True)

In [213]:
pisa_clean['Student ID'].duplicated().sum()

0

In [214]:
pisa_clean.shape

(346570, 19)

<a id='sav'></a>
## Store

In [215]:
# Store the clean DataFrame in a CSV file
pisa_clean.to_csv("pisa_df.csv", index=False, encoding='utf8')

<a id='conclusion'></a>
## Conclusion

<a id='references'></a>
## References